### PR TITLE
Add health and damages

### PR DESCRIPTION
**Pull request by jaumeriera**

### CODE REVIEW

File: Assets/Scripts/Astronaut/AstronautController.cs [hunk 1]
```diff
@@ -16,9 +18,23 @@ public class AstronautController : MonoBehaviour
     private bool pickableObject;
     GameObject objectToPick;
 
-    private void Awake()
+    public void Awake()
```
Comment (Neengash) [inline Assets/Scripts/Astronaut/AstronautController.cs]: private

File: Assets/ScriptableObjects/GoodAlienScriptable.cs [hunk 2]
```diff
@@ -6,9 +6,24 @@ public class GoodAlienScriptable : ScriptableObject
...
 
-    [Range(1,5)]
+    [Range(1,300)]
     public int Health;
 
     [Range(0, 5)]
     public float interactDistance;
+
+    [Range(0, 10)]
+    public int health;
```
Comment (Neengash) [inline Assets/ScriptableObjects/GoodAlienScriptable.cs:25]: duplicated health

File: Assets/ScriptableObjects/Astronaut/AstronautScriptable.cs [hunk 3]
```diff
@@ -5,9 +5,21 @@
 [CreateAssetMenu(fileName = "AstronautScriptable", menuName = "Scriptables/AstronautScriptable", order = 1)]
 public class AstronautScriptable : ScriptableObject
 {
-    [Range(0, 15)]
-    public float velocity;
+    [Range(0, 20)]
+    public float currentVelocity;
```
Comment (Neengash) [inline Assets/ScriptableObjects/Astronaut/AstronautScriptable.cs]: nopes

File: Assets/Scripts/Astronaut/AstronautController.cs [hunk 4]
```diff
@@ -16,9 +18,23 @@ public class AstronautController : MonoBehaviour
...
     {
         visibility = GetComponent<VisibilityInteraction>();
+        healthManager = GetComponent<HealthManager>();
+    }
+
+    private void Start()
+    {
+        visibility.InLight += RestoreVelocity;
+        visibility.OutLight += GoSlower;
```
Comment (Neengash) [inline Assets/Scripts/Astronaut/AstronautController.cs:31]: Unsubscribe OnDestroy

visibility.InLight -= RestoreVelocity;
visibility.OutLight -= GoSlower;

File: Assets/Scripts/Astronaut/AstronautController.cs [hunk 5]
```diff
@@ -16,9 +18,23 @@ public class AstronautController : MonoBehaviour
...
+        InitHealthManager();
+        _astronaut.currentVelocity = _astronaut.normalVelocity;
+    }
+
+    private void InitHealthManager() {
+        healthManager.SetUp(_astronaut.health);
+        healthManager.NoHealth += Die;
```
Comment (Neengash) [inline Assets/Scripts/Astronaut/AstronautController.cs:43]: unsubscribe OnDestroy

File: Assets/Scripts/Astronaut/AstronautController.cs [hunk 6]
```diff
@@ -61,13 +76,31 @@ private void Awake()
...
         objectToPick = null;
     }
+
+    public void RestoreVelocity() {
+        _astronaut.currentVelocity = _astronaut.normalVelocity;
+    }
+
+    public void GoSlower() {
+        _astronaut.currentVelocity= _astronaut.slowVelocity;
+    }
```
Comment (Neengash) [inline Assets/Scripts/Astronaut/AstronautController.cs:100]: Todo

File: Assets/Scripts/Astronaut/AstronautMovement.cs [hunk 7]
```diff
@@ -13,8 +13,8 @@ void Update()
 
     private void UpdatePosition()
     {
-        float positionX = this.transform.position.x + (Input.GetAxisRaw("Horizontal") * _astronaut.velocity * Time.deltaTime);
-        float positionY = this.transform.position.y + (Input.GetAxisRaw("Vertical") * _astronaut.velocity * Time.deltaTime);
+        float positionX = this.transform.position.x + (Input.GetAxisRaw("Horizontal") * _astronaut.currentVelocity * Time.deltaTime);
+        float positionY = this.transform.position.y + (Input.GetAxisRaw("Vertical") * _astronaut.currentVelocity * Time.deltaTime);
```
Comment (Neengash) [inline Assets/Scripts/Astronaut/AstronautMovement.cs]: currentVelocity

File: Assets/Scripts/Astronaut/FlareLauncher/Flare.cs [hunk 8]
```diff
@@ -2,20 +2,32 @@
...
         collider = GetComponent<CircleCollider2D>();
+        InitHealthManager();
     }
 
-    // Update is called once per frame
-    void Update() {
-        
+    private void InitHealthManager() {
+        healthManager.SetUp(_flare.health);
+        healthManager.NoHealth += Die;
```
Comment (Neengash) [inline Assets/Scripts/Astronaut/FlareLauncher/Flare.cs:29]: unsubscribe

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs [hunk 9]
```diff
@@ -29,36 +34,50 @@ public class GoodAlienMain : MonoBehaviour
 
         visibility.InLight += EnterLight;
         visibility.OutLight += ExitLight;
+
+        InitHealthManager();
+    }
+    private void InitHealthManager() {
+        healthManager.SetUp(_alien.health);
+        healthManager.NoHealth += Die;
```
Comment (Neengash) [inline Assets/Scripts/GoodAlien/GoodAlienMain.cs:44]: Unsubscribe

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs [hunk 10]
```diff
@@ -29,36 +34,50 @@ public class GoodAlienMain : MonoBehaviour
...
-        // StartHealing
+        if (healingCoroutineReference != null) {
+            StopCoroutine(healingCoroutineReference);
+        }
     }
     private void ExitLight()
     {
         Debug.Log("ExitLight");
-
+        if (healingCoroutineReference != null) {
```
Comment (Neengash) [inline Assets/Scripts/GoodAlien/GoodAlienMain.cs]: == null

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs [hunk 11]
```diff
@@ -29,36 +34,50 @@ public class GoodAlienMain : MonoBehaviour
...
+            healingCoroutineReference = StartCoroutine(Healing());
+        }
     }
 
     // TODO :: THIS IS TEMPORARY
     void Update() {
         if (Input.GetKeyDown(KeyCode.Space)) {
             SwitchController();
         }
```
Comment (Neengash) [inline Assets/Scripts/GoodAlien/GoodAlienMain.cs]: Remove Update

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs [hunk 12]
```diff
@@ -29,36 +34,50 @@ public class GoodAlienMain : MonoBehaviour
...
     public void SwitchController() {
         freeController.enabled = !freeController.isActiveAndEnabled;
         holdController.enabled = !holdController.isActiveAndEnabled;
     }
 
+    private IEnumerator Healing()
+    {
+        while(healthManager.CanHeal()) {
+            healthManager.Heal(_alien.HealValue);
+            yield return null;
```
Comment (Neengash) [inline Assets/Scripts/GoodAlien/GoodAlienMain.cs]: yield return new WaitForSeconds (scriptableConfiguration)

**Review by Neengash — CHANGES_REQUESTED** [review]

File: Assets/ScriptableObjects/GoodAlienScriptable.cs (hunk 2 above)
Comment (jaumeriera) [reply Assets/ScriptableObjects/GoodAlienScriptable.cs:25]: Removed

File: Assets/Scripts/Astronaut/AstronautController.cs (hunk 1 above)
Comment (jaumeriera) [reply Assets/Scripts/Astronaut/AstronautController.cs]: Done

File: Assets/Scripts/Astronaut/AstronautController.cs (hunk 5 above)
Comment (jaumeriera) [reply Assets/Scripts/Astronaut/AstronautController.cs:43]: Done

File: Assets/Scripts/Astronaut/AstronautController.cs (hunk 6 above)
Comment (jaumeriera) [reply Assets/Scripts/Astronaut/AstronautController.cs:100]: Done

File: Assets/Scripts/Astronaut/AstronautMovement.cs (hunk 7 above)
Comment (jaumeriera) [reply Assets/Scripts/Astronaut/AstronautMovement.cs]: Done

File: Assets/Scripts/Astronaut/FlareLauncher/Flare.cs (hunk 8 above)
Comment (jaumeriera) [reply Assets/Scripts/Astronaut/FlareLauncher/Flare.cs:29]: Done

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs (hunk 9 above)
Comment (jaumeriera) [reply Assets/Scripts/GoodAlien/GoodAlienMain.cs:44]: Done

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs (hunk 10 above)
Comment (jaumeriera) [reply Assets/Scripts/GoodAlien/GoodAlienMain.cs]: done

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs (hunk 11 above)
Comment (jaumeriera) [reply Assets/Scripts/GoodAlien/GoodAlienMain.cs]: done

File: Assets/Scripts/GoodAlien/GoodAlienMain.cs (hunk 12 above)
Comment (jaumeriera) [reply Assets/Scripts/GoodAlien/GoodAlienMain.cs]: done

**Pull request merged**